typing/typetexp.ml

```diff
@@ -472,29 +498,17 @@ and transl_type_aux env policy styp =
   | Ptyp_poly(vars, st) ->
       let vars = List.map (fun v -> v.txt) vars in
       begin_def();
-      let new_univars = List.map (fun name -> name, newvar ~name ()) vars in
+      let new_univars = make_poly_univars vars in
       let old_univars = !univars in
       univars := new_univars @ !univars;
       let cty = transl_type env policy st in
       let ty = cty.ctyp_type in
       univars := old_univars;
       end_def();
       generalize ty;
-      let ty_list =
-        List.fold_left
-          (fun tyl (name, ty1) ->
-            let v = Btype.proxy ty1 in
-            if deep_occur v ty then begin
-              match v.desc with
-                Tvar name when v.level = Btype.generic_level ->
-                  Btype.set_type_desc v (Tunivar name);
-                  v :: tyl
-              | _ ->
-                raise (Error (styp.ptyp_loc, env, Cannot_quantify (name, v)))
-            end else tyl)
-          [] new_univars
-      in
-      let ty' = Btype.newgenty (Tpoly(ty, List.rev ty_list)) in
+      let ty_list = check_poly_univars env styp.ptyp_loc new_univars in
+      let ty_list = List.filter (fun v -> deep_occur v ty) ty_list in
+      let ty' = Btype.newgenty (Tpoly(ty, ty_list)) in
```

Comment:
The result of `deep_occur` can't be affected by any of the code in its old position and its new position, and moving it to its new position makes `check_poly_univars` more reusable.

That means the only possible change in behaviour is from running the check on variables that would have been filtered out. Since such variables don't appear in `ty` they need to be generalized separately in order to pass the check -- that is what the additional call to `generalize` in `check_poly_univars` does. I think that should be sufficient to ensure that all such variables pass the check, but if it somehow isn't then the new code is actually more correct than the old code.